clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -3205,6 +3205,40 @@ void CIRGenModule::applyReplacements() {
     auto NewF = dyn_cast<cir::FuncOp>(Replacement);
     assert(NewF && "not implemented");
 
+    // LLVM has opaque pointer but CIR not. So we may have to handle these
+    // different pointer types when performing replacement.
+    auto optionalUseRange = OldF.getSymbolUses(theModule);
+    if (optionalUseRange) {
+      for (auto U : *optionalUseRange) {
+        auto Call = mlir::dyn_cast<cir::CallOp>(U.getUser());
+        if (!Call)
+          continue;
+
+        auto ArgOps = Call.getArgOps();
+        auto FuncArgTypes = NewF.getFunctionType().getInputs();
+        for (unsigned I = 0; I < FuncArgTypes.size(); I++) {
+          if (ArgOps[I].getType() == FuncArgTypes[I])
+            continue;
+
+          auto argPointerTy =
+              mlir::dyn_cast<cir::PointerType>(ArgOps[I].getType());
+          auto funcArgPointerTy =
+              mlir::dyn_cast<cir::PointerType>(FuncArgTypes[I]);
+
+          // If we can't solve it, leave it for the verifier to bail out.
+          if (!argPointerTy || !funcArgPointerTy)
+            continue;
+
+          mlir::OpBuilder::InsertionGuard guard(builder);
+          builder.setInsertionPoint(Call);
+          auto castedArg =
+              builder.createBitcast(Call.getLoc(), ArgOps[I], funcArgPointerTy);
+          // FIXME: Raw API. Wrap this into cir::CallOp.
+          Call.setOperand(I, castedArg);
```

Comment:
Please add the nice API as part of this PR.

Reply:
Done